src/cargo/util/toml.rs

```diff
@@ -540,6 +540,18 @@ impl TomlManifest {
             }
         }
 
+        {
+            let mut names_sources = HashMap::new();
+            for dep in deps.iter() {
+                let name = dep.name();
+                let sources = names_sources.entry(name).or_insert(HashSet::new());
+                if sources.insert(dep.source_id()) && sources.len() > 1 {
```

Comment:
Couldn't this be `if !sources.insert(source_id)`? I think `insert` returns false if it was already present